actors/miner/src/lib.rs

```diff
@@ -2995,6 +2996,301 @@ impl Actor {
+            ) {
+                return Err(actor_error!(
+                    forbidden,
+                    "cannot move to deadline {} during its challenge window, \
+                        or the prior challenge window,
+                        or before {} epochs have passed since its last challenge window ended",
+                    params.to_deadline,
+                    policy.wpost_dispute_window
+                ));
+            }
+
+            if !deadline_available_for_move(
+                policy,
+                params.from_deadline,
+                params.to_deadline,
+                current_deadline.index,
+            ) {
+                return Err(actor_error!(
+                    forbidden,
+                    "cannot move to a deadline which is further from current deadline"
+                ));
+            }
+
+            let from_quant = state.quant_spec_for_deadline(policy, params.from_deadline);
+            let to_quant = state.quant_spec_for_deadline(policy, params.to_deadline);
+
+            let mut from_deadline =
+                deadlines.load_deadline(policy, store, params.from_deadline).map_err(|e| {
+                    e.downcast_default(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        format!("failed to load deadline {}", params.from_deadline),
+                    )
+                })?;
+            let mut to_deadline =
+                deadlines.load_deadline(policy, store, params.to_deadline).map_err(|e| {
+                    e.downcast_default(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        format!("failed to load deadline {}", params.to_deadline),
+                    )
+                })?;
+
+            let partitions = &mut params.partitions;
+            if partitions.is_empty() {
+                let partitions_amt = from_deadline.partitions_amt(rt.store()).map_err(|e| {
+                    e.downcast_default(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        format!("failed to load partitions for deadline {}", params.from_deadline),
+                    )
+                })?;
+
+                for partition_idx in 0..partitions_amt.count() {
+                    partitions.set(partition_idx);
+                }
+            }
+
+            let (live, dead, removed_power) =
+                from_deadline.remove_partitions(store, partitions, from_quant).map_err(|e| {
+                    e.downcast_default(
+                        ExitCode::USR_ILLEGAL_STATE,
+                        format!(
+                            "failed to remove partitions from deadline {}",
+                            params.from_deadline
+                        ),
+                    )
+                })?;
+
+            state.delete_sectors(store, &dead).map_err(|e| {
+                e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "failed to delete dead sectors")
+            })?;
```

Comment:
I'm looking into the "moving partitions intact" approach, but `Deadline.expirations_epochs` seems quite complicated to maintain while moving.

Reply:
Hmm, I see. I think we should avoid loading sector info if at all possible.

If we say that a partition with faulty sectors can't be moved, then I think we only need to worry about on-time expirations. In that case I suggest rounding up, i.e. find the next epoch quantized to the new deadline that is not sooner than epoch in the queue. Then the SP might need to maintain a sector up to 1 proving period longer than committed.

Reply:
Don't worry about removing anything from the old deadline's queue. It will have redundant entries, but they're harmless.

Reply:
What about other sectors: faulty, unproven, terminated? I was considering to allow all sectors for this "moving partitions intact" approach.

Reply:
I would like to consider allowing any partition to be moved too, but at the moment I don't think it's worth the work/risk of changing how remove_partitions works at the moment. Let's require non-faulty, fully proven sectors for now.

Yes I think we should probably remove that test case. The quantisation is mentioned elsewhere too but I think we can probably remove it, and update associated code to handle arbitrary quantisation. This can be an independent change ahead of what you're doing here – for now just comment out that state assertion.

Reply:
Please do push that commit to this branch - I think it's the right approach.

Reply:
> Also, I found there is a [test case](https://github.com/filecoin-project/builtin-actors/blob/8e82ef71a1858d0530e74e778056dc1b67981587/actors/miner/src/testing.rs#L644) which requires the epoch to be exactly quant.quantize_up(sector.expiration), should we modify this check?

I looked into this and I think we can just make a simple change to make it `epoch >= target`.

Reply:
I changed it a bit more restrictive: `epoch >= target || (epoch-target)%quant.unit ==0`